automotive/maliput/utility/generate_obj.cc

```diff
@@ -158,6 +183,248 @@ class GeoFace {
+//                |     ^                        |
+//                v     |                  +r <--o
+//                o --> *
+//    top_face - v3     top_face - v0
+//
+// bottom_face - v2     bottom_face - v1
+//                o <-- o
+//                |     ^
+//                v     |
+//                o --> *
+// bottom_face - v3      bottom_face - v0
+//
+// Result:
+//
+//    top_face - v2     top_face - v1
+//                o <-- o
+//                |     ^
+//                |     |
+//                v     |
+//                o --> *
+// bottom_face - v3      bottom_face - v0
+//
+// @param bottom_face  the GeoFace at the bottom.
+// @param top_face     the GeoFace at the top.
+GeoFace mergeQuadsInS(const GeoFace& bottom_face, const GeoFace& top_face) {
+  return GeoFace({bottom_face.vertices()[0], top_face.vertices()[1],
+                  top_face.vertices()[2], bottom_face.vertices()[3]},
+                 {bottom_face.normals()[0], top_face.normals()[1],
+                  top_face.normals()[2], bottom_face.normals()[3]});
+}
+
+// Merge as many adjacent quads as possible. Quads are assumed to be ordered
+// from right to left in lane coordinates (i.e. increasing in s)
+//
+// @param quads  the collection of quads we attempt to merge.
+// @return a new collection of quads.
+std::vector<GeoFace> MergeRowQuads(const std::vector<GeoFace>& quads) {
+  if (quads.size() < 2) {
+    return quads;
+  }
+
+  std::vector<GeoFace> merged_quads;
+  GeoFace accumulated = quads[0];
+
+  for (size_t index = 1; index < quads.size(); ++index) {
+    GeoFace current_face = quads[index];
+    if (canMergeQuadsInR(accumulated, current_face)) {
+      accumulated = mergeQuadsInR(accumulated, current_face);
+    } else {
+      merged_quads.push_back(accumulated);
+      accumulated = current_face;
+    }
+  }
+
+  merged_quads.push_back(accumulated);
+
+  return merged_quads;
+}
+
+// Merge as many adjacent quads as possible between two rows. On each quad row,
+// quads are assumed to be ordered from right to left in lane coordinates
+// (i.e. increasing in s). Rows are supposed the be ordered in r (i.e.
+// `bottom_row` should come before `top_row` in terms of r).
+//
+// @param bottom_row  a collection of quads.
+// @param top_row     a collection of quads.
+// @return a par of of new quad collections, where merged quads (if any) only
+//         appear in the new top row.
+std::pair<std::vector<GeoFace>, std::vector<GeoFace>> MergeQuadsAcrossRows(
+    const std::vector<GeoFace>& bottom_row,
+    const std::vector<GeoFace>& top_row) {
+  if (bottom_row.empty()) {
+    return std::make_pair(bottom_row, top_row);
+  } else {
+    std::vector<GeoFace> optimized_bottom_row;
+    std::vector<GeoFace> optimized_top_row;
+
+    size_t bottom_index = 0;
+    size_t top_index = 0;
+
+    // The rightmost vertex of the strip. This vertex is shared by the first
+    // quad of each row.
+    auto start_xyz = top_row[0].vertices()[0].v().xyz();
```

Comment:
`auto` does not apply here, you should use `const Vector3<double>`.

Reply:
Changed